feat: `bind:innerText` for `contenteditable`

himynameisdave opened the pull request:
🎉  Adds **`bind:innerText`**!!! #3311

As stated in the issue:

> There is already **bind:innerHTML** and **bind:textContent** implemented for elements with contenteditable="true". But there is also a third property that is also would nice to have: **bind:innerText**.
> 
> InnerText approximates the text the user would get if they highlighted the contents of the element with the cursor and then copied it to the clipboard

Also adds the following to the docs (which might help you, dear reader, better understand the differences between these):

> Elements with the `contenteditable` attribute support the following bindings:
> - [`innerHTML`](https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML)
> - [`innerText`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/innerText)
> - [`textContent`](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent)
> 
> There are slight differences between each of these, read more about them [here](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent#Differences_from_innerText).

Resolves #3311

himynameisdave commented:
Ready for review! Let me know if I missed anything / if something should be updated!

dkondrad commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -634,19 +635,15 @@ export default class Element extends Node {
 					});
 				}
 			} else if (
```
Fix w/s: 
`else if (isNameContenteditable(name)) {`

dkondrad commented on site/content/tutorial/06-bindings/08-contenteditable-bindings/text.md:
````diff
@@ -2,11 +2,16 @@
 ></div>
-```
\ No newline at end of file
+```
````
Suggestion: 
Update the tutorial so demonstrate the differences instead of just linking to documentation.

himynameisdave replied:
Yes, great idea 👍

dkondrad commented on src/compiler/compile/utils/__test__.ts:
```diff
@@ -9,8 +21,89 @@ describe('get_name_from_filename', () => {
+		});
+	});
 
-	it('handles unusual filenames', () => {
```
These deleted lines need to be restored as part of the `get_name_from_filename` test group.

himynameisdave replied:
See https://github.com/sveltejs/svelte/pull/4262, where I address this issue directly (because this test fails and wasn't even testing what it was initially written for).

dkondrad replied:
Hmm, I don't think the change over on that one is quite right either (I'll add a comment over there).

The issue here is that you changed the same lines in 2 PRs which will create a conflict in whichever one is merged second.  Maybe that's unavoidable in this case.

himynameisdave replied:
Yep there might be a conflict, which I don't mind resolving 🙂 I figured #4262 would probably be merged first anyway, since it's a lot simpler, plus I still need to add runtime test cases and address your other comments in this PR. 👍

dkondrad commented on src/compiler/compile/utils/contenteditable.ts:
```diff
@@ -0,0 +1,60 @@
+import Element from '../nodes/Element';
+
+
+export const CONTENTEDITABLE_ATTR = 'contenteditable';
```
Probably a bit overkill to make the attribute name a const export.

himynameisdave replied:
Yeah generally I would agree with you, but I did this to avoid mis-typing it as `'contentEditable'`, which I did a few times while implementing this PR 😅

Conduitry commented on src/compiler/compile/utils/contenteditable.ts:
```diff
@@ -0,0 +1,60 @@
+ * Returns true if node is an 'input' or 'textarea'.
+ * @param {Element} node The element to be checked
+ */
+export function isInputOrTextarea(node: Element): boolean {
```
Can you switch to `snake_case` please for internal identifiers to match the style of the rest of the project?

fraserdarwent commented:
What's the state of this?

Conduitry commented:
I believe it had unaddressed reviewer comments, and I see that it now also has conflicts with master.

himynameisdave commented:
@fraserdarwent && @Conduitry  I was blocked because I was unsure of how to test edge cases (vs `bind:innerHTML` or `bind:textContent`). I think since originally implementing this, I have a better understanding of those edge cases so I should be able to add proper test cases for those now.

nltesown commented:
Hello, this feature would be very useful for the project I'm currenly working on. If I understand correctly, the implementation is done but there are still some merging conflicts... What's supposed to happen at this point? @Conduitry @himynameisdave, is there something you can do? Thanks!

scripttease commented:
Hello! Would it be possible for this to be completed at some point? Being able to get the text content with the newlines preserved would be very useful, thank you.

himynameisdave commented:
Update: I have rebased, addressed _most_ of the code review feedback (expect [this](https://github.com/sveltejs/svelte/pull/4291#discussion_r368815664)), however I now have some failing tests which I need help addressing (cc: @Conduitry && @dkondrad). Sorry for taking so long to get to this, but I'd love to get it over the finish line soon, so let me know if there is anything I need to update.

tcc-sejohnson commented:
@himynameisdave 

I rebased this [here](https://github.com/tcc-sejohnson/svelte/tree/issues/3311). It is now up-to-date with `master`. I opened a pull request to incorporate those updates into Dave's branch. (We want him to get credit for his hard work!)

Here's the _super_ confusing thing: The `runtime` tests are failing, and I _cannot_ figure out why.

I copied the tests straight from `binding-contenteditable-text` and adapted them. The test's `main.svelte` looks like this:

```js
<script>
	export let name;
</script>

<editor contenteditable="true" bind:innerText={name}></editor>
<p>hello {name}</p>
```

So you would expect it to render this if provided with `world` for the `name` prop:

```html
<editor contenteditable="true">world</editor>
<p>hello world</p>
```

However, what it renders (according to the test output) is this:
```html
<editor contenteditable="true"></editor>
<p>hello world</p>
```

Notice the difference? The binding is _working_ for the `p` element, but it is _not_ working for the `editor` element. There must be _something_ going on where Svelte is overriding `textContent` which is causing that to render nothing inside the `editor` element.

Even weirder, if you look at the compiler output for `binding-contenteditable-innertext` and diff it with the output for `binding-contenteditable-text` (which really should be named `-textcontent`), they are _exactly_ the same, save that `textContent` is replaced with `innerText`. I mean, _exactly_ the same. So why are they behaving differently?!

This also completely works in a live app -- both the initial value and the binding.

I may do some other searching on where in the compiler Svelte is assigning an empty string to `textContent`. That's the only thing I can think of that could possibly be happening.

Indeedornot commented:
Hi. What is the state of this pull request? Found myself looking for it because `bind:textContent` doesn't play nice with `\n`

tcc-sejohnson commented:
@Indeedornot Nothing has happened since my last post.

flbn commented:
updates on this?

dummdidumm reviewed (approved):
Thank you for implementing this!

I found out why the tests failed - it's because JSDom does not support `innerText` yet. I adjusted the test so it still checks some things while working around the limitations.
